### PR TITLE
[front] - feature(api/assistant): conversation & messages rate limit

### PR DESCRIPTION
**Pull request by JulesBelveze**
## Description

- Add a rate limiting mechanism to prevent overuse of resources by setting limits on user message posting and conversation creation per active user within a workspace
- Rate limits are determined by the number of active users, with a specified maximum number of actions allowed per timeframe

## Deploy Plan

Deploy `front`

### CODE REVIEW

File: front/lib/api/assistant/conversation.ts [hunk 1]
```diff
@@ -584,6 +584,31 @@ export async function* postUserMessage(
   const subscription = auth.subscription();
   const plan = subscription?.plan;
 
+  const activeUsers = await countActiveSeatsInWorkspaceCached(
```
Comment (spolu) [inline front/lib/api/assistant/conversation.ts:587]: Should come after the owner check below

Comment (spolu) [reply front/lib/api/assistant/conversation.ts:587]: Actually let's move rate limits in API endpoint. Repeats a bit of code but really this is the endpoint responsibility not the internal lib 🙏

File: front/pages/api/v1/w/[wId]/assistant/conversations/index.ts [hunk 2]
```diff
@@ -63,6 +67,26 @@ async function handler(
 
   switch (req.method) {
     case "POST":
+      const activeUsers = await countActiveSeatsInWorkspaceCached(owner.sId);
+      const maxPerTimeframe = 10 * activeUsers;
+
+      const remaining = await rateLimiter({
+        key: `createConversation:${auth.workspace()?.id}`,
```
Comment (spolu) [inline front/pages/api/v1/w/[wId]/assistant/conversations/index.ts:74]: Wrong id

Comment (spolu) [reply front/pages/api/v1/w/[wId]/assistant/conversations/index.ts:74]: As it's not aligned with the other (model is vs sid)

**Review by spolu — CHANGES_REQUESTED** [review]

**Review by spolu — COMMENTED** [review]
Sorry reviewing the code again it seems we do have a function that is already in lib/api/conversation and already pulls the activeSeats

This will save us one call to Redis.

Let's move the logic to `isMessagesLimitReached`

We don't need to account for each mentions here it's more a message thing

**Comment by spolu** [comment]
Sorry for the back and fort :see_no_evil:

**Comment by JulesBelveze** [comment]
My fault, too much context switching I'm making stupid mistakes 😓

**Comment by JulesBelveze** [comment]
@spolu hope I understood correctly

File: front/lib/api/assistant/conversation.ts [hunk 3]
```diff
@@ -595,19 +594,18 @@ export async function* postUserMessage(
...
-    mentions,
   });
   if (isAboveMessageLimit) {
     yield {
       type: "user_message_error",
       created: Date.now(),
       error: {
-        code: "plan_message_limit_exceeded",
-        message: "The message limit for this plan has been exceeded.",
+        code: "rate_limit_error",
```
Comment (spolu) [inline front/lib/api/assistant/conversation.ts:607]: This is a user visible error. We likely want to have two messages based on what limit was reached. We can return a boolean and a type of limit exceeded and have 2 different messages here

**Review by spolu — CHANGES_REQUESTED** [review]

**Comment by spolu** [comment]
Happy to chat IRL if helpful

**Comment by spolu** [comment]
np

**Comment by JulesBelveze** [comment]
@spolu Do the new changes reflect what you have in mind?

**Review by spolu — APPROVED** [review]
LGTM

**Pull request merged**